File: node/src/main/java/dev.httpmarco.polocloud.node/commands/type/IntArgument.java
package dev.httpmarco.polocloud.node.commands.type;

import dev.httpmarco.polocloud.node.commands.CommandArgument;
import dev.httpmarco.polocloud.node.commands.CommandContext;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Unmodifiable;

import java.util.List;

public final class IntArgument extends CommandArgument<Integer> {

    public IntArgument(String key) {
        super(key);
    }

    @Override
    public boolean predication(@NotNull String rawInput) {
        try {
            Integer.parseInt(rawInput);
            return true;
        } catch (NumberFormatException e) {
            return false;
        }
    }

    @Contract(pure = true)
    @Override
    public @NotNull String wrongReason() {
        return "The argument " + key() + " is not a number!";
    }

    @Contract(pure = true)
    @Override
    public @Unmodifiable List<String> defaultArgs(CommandContext context) {
        return List.of();
    }

    @Contract(pure = true)
    @Override
    public @NotNull Integer buildResult(String input) {
        return Integer.parseInt(input);
    }
}

File: agent/src/main/kotlin/dev/httpmarco/polocloud/agent/runtime/local/terminal/arguments/type/IntArgument.kt
package dev.httpmarco.polocloud.agent.runtime.local.terminal.arguments.type

import dev.httpmarco.polocloud.agent.i18n
import dev.httpmarco.polocloud.agent.runtime.local.terminal.arguments.TerminalArgument
import dev.httpmarco.polocloud.agent.runtime.local.terminal.arguments.InputContext

class IntArgument(key: String, val minValue: Int? = null, val maxValue: Int? = null) : TerminalArgument<Int>(key) {
    override fun predication(rawInput: String): Boolean {
        return getError(rawInput) == null
    }

    override fun wrongReason(rawInput: String): String {
        return when (getError(rawInput)) {
            IntArgumentError.INVALID_INT -> i18n.get("agent.terminal.command.argument.type.int.wrong")
            IntArgumentError.LOWER_THAN_MIN -> i18n.get(
                "agent.terminal.command.argument.type.int.wrong.atleast",
                key,
                minValue
            )

            IntArgumentError.HIGHER_THAN_MAX -> i18n.get(
                "agent.terminal.command.argument.type.int.wrong.notexceed",
                key,
                maxValue
            )
            else -> ""
        }
    }

    override fun defaultArgs(context: InputContext): MutableList<String> {
        return mutableListOf()
    }

    override fun buildResult(input: String, context: InputContext): Int {
        return input.toInt()
    }

    private fun getError(rawInput: String): IntArgumentError? {
        try {
            val intValue = rawInput.toInt()

            if (maxValue != null && intValue > maxValue) {
                return IntArgumentError.HIGHER_THAN_MAX
            }

            if (minValue != null && intValue < minValue) {
                return IntArgumentError.LOWER_THAN_MIN
            }

            return null
        } catch (_: NumberFormatException) {
            return IntArgumentError.INVALID_INT
        }
    }
}

enum class IntArgumentError {
    INVALID_INT,
    LOWER_THAN_MIN,
    HIGHER_THAN_MAX,
}
